azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,698 @@
+            else:
+                cgroup_path = os.path.join(path_maker(hierarchy), self.name)
+            if not os.path.exists(cgroup_path):
+                logger.info("Creating cgroup directory {0}".format(cgroup_path))
+                CGroups._try_mkdir(cgroup_path)
+            self.cgroups[hierarchy] = cgroup_path
+
+    @staticmethod
+    def is_systemd_manager():
+        """
+        Determine if systemd is managing system services. Many extensions are structured as a set of services,
+        including the agent itself; systemd expects those services to remain in the cgroups in which it placed them.
+        If this process (presumed to be the agent) is in a cgroup that looks like one created by systemd, we can
+        assume systemd is in use.
+
+        :return: True if systemd is managing system services
+        :rtype: Bool
+        """
+        if CGroups._use_systemd is None:
+            hierarchy = HIERARCHIES[0]
+            path = CGroups.get_my_cgroup_folder(hierarchy)
+            CGroups._use_systemd = path.startswith(CGroups.construct_systemd_path_for_hierarchy(hierarchy))
+        return CGroups._use_systemd
+
+    @staticmethod
+    def _try_mkdir(path):
+        """
+        Try to create a directory. If it already exists as such, do nothing. Raise appropriate exceptions if an error
+        should occur.
+
+        :param path: str
+        """
+        if not os.path.isdir(path):
+            try:
+                os.mkdir(path)
+            except OSError as e:
+                if e.errno == errno.EEXIST:
+                    if not os.path.isdir(path):
+                        raise CGroupsException(
+                            "Create directory for cgroup {0}: normal file already exists with that name".format(path)
+                        )
+                    else:
+                        pass    # There was a race to create the directory, but it's there now, and that's fine
+                elif e.errno == errno.EACCES:
+                    # This is unexpected, as the agent runs as root
+                    raise CGroupsException("Create directory for cgroup {0}: permission denied".format(path))
+                else:
+                    raise OSError(e)
+
+    @staticmethod
+    def construct_custom_path_for_hierarchy(hierarchy):
+        return os.path.join(BASE_CGROUPS, hierarchy, AGENT_NAME)
+
+    @staticmethod
+    def construct_systemd_path_for_hierarchy(hierarchy):
+        return os.path.join(BASE_CGROUPS, hierarchy, 'system.slice')
+
+    def add(self, pid):
+        """
+        Add a process to the cgroups for this agent/extension.
+        """
+        if not self.enabled():
+            return
+
+        if self.is_wrapper_cgroup:
+            raise CGroupsException("Cannot add a process to the Agents+Extensions wrapper cgroup")
+
+        try:
+            # determine if pid exists by sending signal 0 to it
+            os.kill(pid, 0)
```

Comment:
This should just be a method in osutil.  This will avoid the need for a comment in the first place, and is useful in other parts of the code.  You can do it now, or open an issue to track.

```python
  def is_process_alive(pid):
      try:
          os.kill(pid, 0)
          return True
      except OSError:
          return False
```

Reply:
There already is check_pid_alive(), but it's a worse implementation (calls os.path.isdir('/proc/<pid>')). I'll fix that and use it.